sdxl-lora-swapping/model/model.py

```diff
@@ -0,0 +1,96 @@
+from typing import Any
+
+from diffusers import DiffusionPipeline, AutoencoderKL
+import torch
+import base64
+from io import BytesIO
+from PIL import Image
+import time
+import functools
+
+
+# Good notebook to learn how to use diffusers + LoRA:
+# https://colab.research.google.com/gist/sayakpaul/109b7e792c64514fb3c057ecff4145ff/scratchpad.ipynb#scrollTo=GreOMxAcvlm8
+
+class Model:
+    def __init__(self, **kwargs) -> None:
+        self.pipe = None
+
+    def load(self):
+        # Load model here and assign to self._model.
+        vae = AutoencoderKL.from_pretrained(
+            "madebyollin/sdxl-vae-fp16-fix", torch_dtype=torch.float16
+        )
+        self.pipe = DiffusionPipeline.from_pretrained(
+            "stabilityai/stable-diffusion-xl-base-1.0",
+            vae=vae,
+            torch_dtype=torch.float16,
+            variant="fp16",
+            use_safetensors=True,
+        )
+
+        self.pipe.load_lora_weights(
+            "stabilityai/stable-diffusion-xl-base-1.0",
+            weight_name="sd_xl_offset_example-lora_1.0.safetensors",
+        )
+         
+        self.pipe.to('cuda')
+        
+        self.refiner = DiffusionPipeline.from_pretrained(
+            "stabilityai/stable-diffusion-xl-refiner-1.0",
+            text_encoder_2=self.pipe.text_encoder_2,
+            vae=self.pipe.vae,
+            torch_dtype=torch.float16,
+            use_safetensors=True,
+            variant="fp16",
+        )
+        self.refiner.to("cuda")
+
+    def convert_to_b64(self, image: Image) -> str:
+        buffered = BytesIO()
+        image.save(buffered, format="JPEG")
+        img_b64 = base64.b64encode(buffered.getvalue()).decode("utf-8")
+        return img_b64
+
+    def predict(self, model_input: Any) -> Any:
+        prompt = model_input.pop("prompt")
+        negative_prompt = model_input.pop("negative_prompt", None)
+        target_size = model_input.pop("size", 1024)
+        use_refiner = model_input.pop("use_refiner", True)
+        high_noise_frac = model_input.pop("high_noise_frac", 0.8)
+        num_inference_steps = model_input.pop("num_inference_steps", 30)
+
+        lora = model_input.pop("lora", None)
+        print(f"Loading LoRA weights: {lora}")
+
+        # example schema: 
+        # {"repo_id": "nerijs/pixel-art-xl", "weights": "pixel-art-xl.safetensors"}
+
+        if lora is not None:
```

Comment:
Does this always download from remote if `lora` kwarg is in the input? What if the correct adapter is already loaded in gpu memory from previous request?

Reply:
It won't redownload if it was used last. The default behavior here is to use the last loaded LoRA since it modifies the `pipe` itself.